Introduce tropical geometry

lkastner opened the pull request:
This is a subset of the big tropical geometry branch which contains the basic architecture, but does not need the `pAdicSolver`. So this can be merged and then some other branches will be rebased onto this.

YueRen committed:
TropicalGeometry: initial commit

lkastner committed:
TropicalGeometry: Make TropicalHypersurface(poly) work

lkastner committed:
TropicalGeometry: Homogenize inhomogeneous input for polymake

lkastner committed:
TropicalGeomtry: Ask for vertices of tropical hypersurface

YueRen committed:
chg: moving files around, adding basic properties for tropical varieties

YueRen committed:
fix: some broken properties

YueRen committed:
fix: broken property

lkastner committed:
TropicalGeometry: Add stably_intersect

lkastner committed:
TropicalGeometry: Also homogenize homogeneous polynomials, i.e. add variable

YueRen committed:
fix: tropical numbers

lkastner committed:
TropicalGeometry: Remove brainstorming leftovers

lkastner committed:
TropicalGeometry: Nicer show methods

lkastner committed:
TropicalGeometry: intersect tropical varieties

lkastner committed:
TropicalGeometry: Add documentation to intersect function

lkastner committed:
TropicalGeometry: Fix docstring for new Base.show

lkastner committed:
TropicalGeometry: Eliminate most warnings from duplicate definitions

YueRen committed:
chg (broken): restructured codebase

lkastner committed:
TropicalGeometry: Add constructor according to Note in AbstractAlgebra.jl

lkastner committed:
TropicalGeometry: Start documentation

- Add reference to Sturmfels-MacLagan book
- Embed in Algebraic Geometry section

lkastner committed:
PolyhedralGeometry: Access to basic properties of PolyhedralComplex

YueRen committed:
chg (broken): constructors for tropical hypersurfaces

YueRen committed:
fix: hypersurface from algebraic polynomial

YueRen committed:
fix: tropical_polynomial coefficients depend on min-max convention

lkastner committed:
TropicalGeometry: Extract underlying poyhedral complex

YueRen committed:
chg: adding more documentation

YueRen committed:
chg: initial changes to base everything on PolyhedralComplex

YueRen committed:
fix: various bugs

lkastner committed:
PolyhedralGeometry: Add some more basic functionality for PolyhedralComplex

YueRen committed:
TropicalGeometry: initial commit for valuation.jl

YueRen committed:
TropicalGeometry: fixing initial forms to work with ValuationMap

YueRen committed:
TropicalGeometry: support for trivial valuations

YueRen committed:
TropicalGeometry: initial commit for initial.jl

YueRen committed:
TropicalGeometry: initial commit for groebner_basis.jl

YueRen committed:
TropicalGeometry: initial (broken) version for tropical groebner bases

YueRen committed:
TropicalGeometry: fixing tropical groebner bases

YueRen committed:
TropicalGeometry: fix groebner_basis header

YueRen committed:
TropicalGeometry: sanatizing input and names in groebner_bases.jl

YueRen committed:
TropicalGeometry: adding complete_reduction parameter to groebner basis

YueRen committed:
TropicalGeometry: using MPolyBuildCtx in pseudo_change_base_ring

YueRen committed:
TropicalGeometry: desimulate_valuation

YueRen committed:
TropicalGeometry: code cleanup

- moved simulate_valuation and desimulate_valuation from groebner_basis.jl to valuation.jl
- merged pseudo_change_base_ring into simulate_valuation

YueRen committed:
TropicalGeometry: added return_lead parameter to groebner_basis

YueRen committed:
TropicalGeometry: initial commit groebner_polyhedron.jl

YueRen committed:
TropicalGeometry: new homogeneity_space

YueRen committed:
TropicalGeometry: code cleanup using Iterators.peel

YueRen committed:
TropicalGeometry: initial commit for points.jl, added random_linear_polynomials

YueRen committed:
TropicalGeometry: broken commit so we can continue working on Avi's desktop

YueRen committed:
TropicalGeometry: fixing random_linear_polynomials to return non-homogeous polynomials

YueRen committed:
TropicalGeometry: renamed random_linear_polynomial to random_affine_linear_polynomial

YueRen committed:
TropicalGeometry: first version of tropical_link

YueRen committed:
TropicalGeometry: typos and redundant Oscar.

YueRen committed:
TropicalGeometry: fixed tropical_link to ignore multiplicities

YueRen committed:
TropicalGeometry: fixed check_weight_vector_for_legality

YueRen committed:
TropicalGeometry: code cleanup

YueRen committed:
TropicalGeometry: added desimulate_valuation for weight vectors

YueRen committed:
TropicalGeometry: added docu for groebner_basis

YueRen committed:
TropicalGeometry: fixing valuations + initial for ideals

YueRen committed:
TropicalGeometry: major ValuationMap overhaul

YueRen committed:
TropicalGeometry: added desimulate_valuation for pairs of vectors

YueRen committed:
TropicalGeometry: fixed computing tropical reduced GB

YueRen committed:
TropicalGeometry: fixed tighten_simulation to create distinct x-monomials

YueRen committed:
TropicalGeometry: added show functions for ValuationMap

YueRen committed:
TropicalGeometry: tighten_simulation fixed to divide by gcd of coefficients

YueRen committed:
TropicalGeometry: initial restructured to allow polys, vectors and ideals as input

YueRen committed:
TropicalGeometry: simulate_valuation restructured to allow for vectors of polys as input

Commit:
TropicalGeometry: added weight for trop hypersurfaces

YueRen committed:
TropicalGeometry: is_valuation_*, making tests more flexible

YueRen committed:
TropicalGeometry: tighten_valuation, improved docu

YueRen committed:
TropicalGeometry: t_adic_valuation, extended to more input types

YueRen committed:
TropicalGeometry: TropicalHypersurface, default to min and allow passing valuation as argument

lkastner committed:
TropicalGeometry: renamed tropical_numbers to tropical_semiring

lkastner committed:
docs: Eliminate left-over conflict

lkastner commented:
@thofma There are some changes in the `experimental` section. Is this something you did with @YueRen ?

thofma commented:
No, these are changes/commits that are already merged in the master branch I believe.

thofma commented on src/TropicalGeometry/groebner_basis.jl:
```diff
@@ -0,0 +1,171 @@
+return true if f is homogeneous (w.r.t. total degree)
+return false otherwise
+=======#
+function is_homogeneous(f::Union{AbstractAlgebra.Generic.MPoly{K},fmpq_mpoly,fmpz_mpoly} where {K})
```
There is a reason that we don't have `is[_]homogenous` for arbitrary multivariate polynomials. How does introducing this function work together with the commutative algebra design @wdecker? Should this be added or renamed to an internal function?

fingolfin replied:
perhaps we want such a method (with a single argument) which explicitly uses the total degree as grading. But for now I think it might be best to rename this to e.g. `_is_homogeneous` just to stop people from accidentally relying on it.

lkastner replied:
Ping @YueRen

wdecker replied:
I think renaming for now is a good idea; also, for ideals, is_homogeneous is misleading as it only tests the generators. Can we find a better name for ideals? I will provide an is_homogeneous function  in full generality (for all gradings) soon.

lkastner replied:
I renamed it now to `_is_homogeneous`. I also renamed the other `is_homogeneous` for ideals to `has_homogeneous_generators`.

thofma commented on src/TropicalGeometry/hypersurface.jl:
```diff
@@ -0,0 +1,256 @@
+
+julia> TropicalHypersurface{min}(f)
+
+julia> TropicalHypersurface{max}(f)
```
```suggestion
julia> TropicalHypersurface(f, min)

julia> TropicalHypersurface(f, max)
```
?

lkastner committed:
TropicalGeometry: renamed tropical_numbers to tropical_semiring

lkastner commented:
> No, these are changes/commits that are already merged in the master branch I believe.

Ah, ok, so this one needs some rebasing onto master too. Thanks.

thofma commented on src/TropicalGeometry/hypersurface.jl:
````diff
@@ -0,0 +1,256 @@
+```
+"""
+function TropicalHypersurface(f::Union{AbstractAlgebra.Generic.MPoly{Oscar.TropicalSemiringElem{typeof(min)}},
+                                       AbstractAlgebra.Generic.MPoly{Oscar.TropicalSemiringElem{typeof(max)}}})
````
```suggestion
function TropicalHypersurface(f::MPolyElem{TropicalSemiringElem})
```

thofma reviewed (commented):
Just some random comments.

fingolfin commented on docs/src/TropicalGeometry/intro.md:
```diff
@@ -0,0 +1,29 @@
+
+## Introduction
+
+For notation we refer to [MS15](@cite)
```
```suggestion
For notation we refer to [MS15](@cite).
```

fingolfin commented on src/TropicalGeometry/curve.jl:
```diff
@@ -0,0 +1,67 @@
+
+"""
+function TropicalCurve{M, EMB}(Vertices::Union{SubObjectIterator{<:RayVector}, Oscar.MatElem, AbstractMatrix}, LS::Union{Oscar.MatElem, AbstractMatrix}, Incidence::Matrix{Bool}) where {M, EMB}
+    if EMB
```
Maybe this should say
```suggestion
    error("TODO: not yet implemented")
    if EMB
```

fingolfin commented on src/TropicalGeometry/groebner_basis.jl:
```diff
@@ -0,0 +1,255 @@
+return true if f is homogeneous (w.r.t. total degree)
+return false otherwise
+=======#
+function is_homogeneous(f::Union{AbstractAlgebra.Generic.MPoly{K},fmpq_mpoly,fmpz_mpoly} where {K})
```
The `where` is redundant, so you can write:
```suggestion
function is_homogeneous(f::Union{AbstractAlgebra.Generic.MPoly,fmpq_mpoly,fmpz_mpoly})
```
But actually... perhaps just take the common supertype?
```suggestion
function is_homogeneous(f::MPolyElem)
```

lkastner committed:
TropicalGeometry: Fixing jldoctest's

fingolfin commented on src/TropicalGeometry/groebner_basis.jl:
```diff
@@ -0,0 +1,255 @@
+      return false
+    end
+  end
+  return true
```
More idiomatic would be:
```suggestion
  return all(is_homogeneous, gens(I))
```

fingolfin commented on src/TropicalGeometry/groebner_basis.jl:
```diff
@@ -0,0 +1,255 @@
+@doc Markdown.doc"""
+    groebner_basis(I::Ideal, val::ValuationMap, w::Vector; complete_reduction::Bool, return_lead)
+
+Computes a Groebner basis of `I` over a field with valuation `val` with respect to weight vector `w`, that is a finite generating set of `I` whose initial forms generate the initial ideal with respect to `w`.
```
Please wrap lines, at something between 80-100 columns; otherwise reading these in the Github webview is a pain. Down the road, it can also make it a bit easier to spot changes in there in future diffs

fingolfin commented on src/TropicalGeometry/groebner_basis.jl:
```diff
@@ -0,0 +1,255 @@
+
+Computes a Groebner basis of `I` over a field with valuation `val` with respect to weight vector `w`, that is a finite generating set of `I` whose initial forms generate the initial ideal with respect to `w`.
+
+For the definitions of initial form, initial ideal and Groebner basis see [Maclagan-Sturmfels, Section 2.4].
```
```suggestion
For the definitions of initial form, initial ideal and Groebner basis see Section 2.4 in [MS15](@cite).
```

fingolfin commented on src/TropicalGeometry/groebner_basis.jl:
```diff
@@ -0,0 +1,255 @@
+
+julia> w = [0,0,0,0,0,0];
+
+julia> groebner_basis(Cyclic5Homogenized, val_2, w, complete_reduction=true)
```
Do we really want to print these Gröber bases here?

lkastner replied:
@YueRen I wondered about that too. Do we even need to test this in OSCAR? It definitely makes the documentation harder to read. And I would suspect that these things are already tested inside Singular? Otherwise we should move this to be a regular test instead of a jldoctest.

lkastner replied:
I moved these to the tests. @YueRen will have to add new examples.

fingolfin commented on src/TropicalGeometry/hypersurface.jl:
```diff
@@ -0,0 +1,258 @@
+###
+
+@doc Markdown.doc"""
+    TropicalHypersurface()
```
```suggestion
    TropicalHypersurface(f::MPolyElem{TropicalSemiringElem})
```

fingolfin commented on src/TropicalGeometry/hypersurface.jl:
```diff
@@ -0,0 +1,258 @@
+
+@doc Markdown.doc"""
+    TropicalHypersurface{M}(f::Union{AbstractAlgebra.Generic.MPoly{Oscar.TropicalSemiringElem{typeof(min)}},
+                                     AbstractAlgebra.Generic.MPoly{Oscar.TropicalSemiringElem{typeof(max)}}})
```
This should match the actual function signature.

fingolfin commented on src/TropicalGeometry/semiring.jl:
```diff
@@ -0,0 +1,504 @@
+
+# We use T to record whether we are in the min/max case
+# T is either typeof(min) or typeof(max)
+mutable struct TropicalSemiring{T} <: Field
```
Is there a reason this is `mutable`? Otherwise, just drop it, then it becomes a singleton type, which may enable further optimizations.
```suggestion
struct TropicalSemiring{T} <: Field
```

fingolfin commented on src/TropicalGeometry/semiring.jl:
```diff
@@ -0,0 +1,504 @@
+#   I have no preference, except that it should be consistent with the other libraries,
+#   e.g. what are elements of p-adic number rings called?
+mutable struct TropicalSemiringElem{T} <: FieldElem
+  parent::TropicalSemiring{T}
```
There is no need to store the parent if the parent is a singleton (type) determined by `T`... :-)

fingolfin commented on src/TropicalGeometry/semiring.jl:
```diff
@@ -0,0 +1,504 @@
+# Test if something is inf.
+isinf(x::TropicalSemiringElem) = x.isinf
+
+Oscar.parent(x::TropicalSemiringElem) = x.parent
```
If `TropicalSemiring{T}` is a singleton, then `TropicalSemiring{T}()` always returns the same (identical) object and is also efficient.
```suggestion
Oscar.parent(x::TropicalSemiringElem{T}) where T = TropicalSemiring{T}()
```

fingolfin commented on src/TropicalGeometry/curve.jl:
```diff
@@ -0,0 +1,67 @@
+###
+# 1. Definition
+# -------------
+# M = typeof(min) or typeof(max):
+#   min or max convention, affecting initial ideals
```
While working with `typeof(min)` and `typeof(max)` is cute, it may (!) also be inefficient. It may also be efficient, I've not tested yet alone analyzed this code closer; but before we had the current attribute storing system which uses symbols as dictionary keys, there were some attempts to use functions / function types as dictionary keys, and that was super inefficient. Of course that's a bit of a different situation, as there e.g. hashing is needed for the keys.

But I'd still be vary here... It *might* be better to use symbols (`:min` resp. `:max`) here. Or perhaps [an enum type](https://docs.julialang.org/en/v1/base/base/#Base.Enums.@enum)?

Then again, I might just be overly cautious... I didn't test it here, after all!

@thofma might know better

thofma replied:
I wrote this version with `typeof(min)` and `typeof(max)`. Can you point out where or why this is inefficient? It is the same as storing functions in types a la 
```
struct A{FunctionType}
  f::FunctionType
```
which is pretty common and performant julia code if I understand correctly.

fingolfin replied:
It was slow to use them as keys in a dictionary, mostly because hashing was slower than that of symbols by an order of magnitude. So this may be irrelevant here! This experience simply made me wary. Just for reference:
```
julia> using BenchmarkTools; T = typeof(sin); @btime hash(T); @btime hash(:sin);
  27.945 ns (1 allocation: 16 bytes)
  1.458 ns (0 allocations: 0 bytes)
```

fingolfin replied:
Using them as parameters in parametric types of course is common and should indeed be no issue. Also no problem comparing the functions. Comparing the types of the functions *is* a bit slower, but that should be fine.
```
julia> using BenchmarkTools

julia> @btime min == max; @btime min === max;
  0.833 ns (0 allocations: 0 bytes)
  0.875 ns (0 allocations: 0 bytes)

julia> @btime :min == :max; @btime :min === :max;
  0.875 ns (0 allocations: 0 bytes)
  0.875 ns (0 allocations: 0 bytes)

julia> T = typeof(min); T2=typeof(max) ; @btime T == T2; @btime T === T2;
  55.498 ns (0 allocations: 0 bytes)
  3.625 ns (0 allocations: 0 bytes)

```

Anyway, if @thofma thinks this is fine, I am happy with it :-)

thofma replied:
Everything with type parameters comparisons will hopefully be compiled away. Some of the things are not written optimally, there should be more `===` and `isa`.

fingolfin commented on src/TropicalGeometry/hypersurface.jl:
```diff
@@ -0,0 +1,258 @@
+    if has_attribute(T,:polymake_bigobject)
+        return get_attribute(T,:polymake_bigobject)
+    end
+    error("pm_object(T::TropicalHypersurface): no polymake bigobject attributed")
```
That's a strange of the word "attributed" here (and in several similar places). How about "set" or "provided" or "defined" instead?

fingolfin commented on src/TropicalGeometry/hypersurface.jl:
```diff
@@ -0,0 +1,258 @@
+@doc Markdown.doc"""
+    polynomial(TH::TropicalHypersurface{M, EMB})
+
+Returns the tropical polynomial of `TH` if it is embedded. Returns error otherwise
```
Errors are not returned, they are thrown or raised.

fingolfin commented on src/TropicalGeometry/linear_space.jl:
```diff
@@ -0,0 +1,60 @@
+# Examples
+"""
+function TropicalLinearSpace(ideal::MPolyIdeal{fmpq_poly})
+  return #...
```
I think it would be good if those incomplete functions all had something like `error("TODO")` in them.

But OK, if you are going to fill them very soon, maybe not... But if it is going to take more than a few weeks, perhaps better to fill them in, lest people try to use them and wonder?

YueRen replied:
Good suggestion, should definitely be done. Even if they will be filled in soon, I don't think incomplete functions should be merged into the main branch unless they throw an error when called.

fingolfin commented on src/TropicalGeometry/main.jl:
```diff
@@ -0,0 +1,80 @@
+@attributes _whyohwhy2
+
+_whyohwhy3 = FlintRationalField
+@attributes _whyohwhy3
```
What's this about?

fingolfin commented on src/TropicalGeometry/main.jl:
```diff
@@ -0,0 +1,80 @@
+  end
+
+  return F_gcd
+end
```
```suggestion
gcd(F::Vector{fmpq_poly}) = reduce(gcd, F)
```
In fact, perhaps also allow `gcd(f1,f2,f3,...)`:
```suggestion
gcd(F::Vector{fmpq_poly}) = reduce(gcd, F)
gcd(F::fmpq_poly...) = reduce(gcd, F)
```

thofma replied:
I am not a big fan of these ad hoc definitions here.

fingolfin commented on src/TropicalGeometry/numbers.jl:
```diff
@@ -0,0 +1,449 @@
+
+export tropical_numbers,
+       @tropical,
+       determinant,
```
but it's called `det` everywhere else... if we do this, then I think it should at least be an alias, not just an arbitrary subset of all possible determinant methods. So:
```
const determinant = det
```

lkastner replied:
I renamed it to `det` in order to have some progress.

fingolfin commented on src/TropicalGeometry/numbers.jl:
```diff
@@ -0,0 +1,449 @@
+
+# We use T to record whether we are in the min/max case
+# T is either typeof(min) or typeof(max)
+mutable struct TropicalNumbers{T} <: Field
```
As above, this could be turned into a singleton type.
```suggestion
struct TropicalNumbers{T} <: Field
```

fingolfin replied:
Alas, isn't this and `TropicalSemiring` the same thing, just with a different name?!

lkastner replied:
Apparently `numbers.jl` was just a left-over and not needed anymore. It has been removed.

fingolfin commented on src/TropicalGeometry/semiring.jl:
```diff
@@ -0,0 +1,504 @@
+    return false # x=-inf, no y is smaller
+  end
+  if isinf(y)
+    return !isinf(x) # y=-inf, smaller than all x except x=-inf
```
```suggestion
    return true # y=-inf, smaller than all x except x=-inf, which was handled above
```

fingolfin commented on src/TropicalGeometry/semiring.jl:
```diff
@@ -0,0 +1,504 @@
+    return false # x=inf, no y is smaller
+  end
+  if isinf(y)
+    return !isinf(x) # y=inf, smaller than all x except x=inf
```
```suggestion
    return true # y=inf, smaller than all x except x=inf, which was handled above
```

fingolfin commented on src/TropicalGeometry/semiring.jl:
```diff
@@ -0,0 +1,504 @@
+    else
+      return parent(x)(convention(parent(x))(data(x), data(y)))
+    end
+  end
```
Style remark: I would recommend unnecessary nesting. So:
```suggestion
  if isinf(x)
    return deepcopy(y)
  elseif isinf(y)
    return deepcopy(x)
  else
    return parent(x)(convention(parent(x))(data(x), data(y)))
  end
```
or even
```suggestion
  isinf(x) && return deepcopy(y)
  isinf(y) && return deepcopy(x)
  return parent(x)(convention(parent(x))(data(x), data(y)))
```

fingolfin reviewed (dismissed):
Some feedback. Most of it is stylistic and not super important. But the major code duplication between TropicalSemiring and TropicalNumber should be resolved

lkastner committed:
Apply suggestions from code review

Co-authored-by: Max Horn <max@quendi.de>
Co-authored-by: Tommy Hofmann <thofma@gmail.com>

fieker commented:
On Fri, Mar 11, 2022 at 12:42:43AM -0800, Tommy Hofmann wrote:
> @thofma commented on this pull request.
> 
> 
> 
> > +function gcd(F::Vector{fmpq_poly})
> +  F_gcd,F_latter = Iterators.peel(F)
> +
> +  for f in F_latter
> +    F_gcd = gcd(F_gcd,f)
> +  end
> +
> +  return F_gcd
> +end
> 
> I am not a big fan of these ad hoc definitions here.
... and its inefficient ...
I'd use reduce(gcd, F) instead. Not any better wrt efficiency, but works
now
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/1152#discussion_r824499446
> You are receiving this because you are subscribed to this thread.
> 
> Message ID: ***@***.***>

lkastner committed:
TropicalGeometry: Wrap some lines, improve code

lkastner committed:
TropicalGeometry: Clean up error messages and doc headers

lkastner committed:
TropicalGeometry: Move excessive doctests to test files

lkastner committed:
TropicalGeometry: Move excessive initial tests to test files

lkastner committed:
TropicalGeometry: ValuationMap->TropicalSemiringMap

lkastner committed:
TropicalGeometry: determinant->det

lkastner committed:
TropicalGeometry: Better names for homogeneity checks

YueRen committed:
TropicalGeometry: updated src/TropicalGeometry/hypersurface.jl based on Tommy's suggestion

https://github.com/oscar-system/Oscar.jl/pull/1152#pullrequestreview-904223727

Co-authored-by: Tommy Hofmann <thofma@gmail.com>

fieker commented:
On Wed, Mar 16, 2022 at 05:40:59AM -0700, YueRen wrote:
> @YueRen commented on this pull request.
> 
> 
> 
> > +
> +
> +###
> +# 2. Basic constructors
> +# ---------------------
> +###
> +
> ***@***.*** Markdown.doc"""
> +    TropicalLinearSpace()
> +
> +Construct a tropical linear space from a degree 1 polynomial ideal
> +
> +# Examples
> +"""
> +function TropicalLinearSpace(ideal::MPolyIdeal{fmpq_poly})
> +  return #...
> 
> Good suggestion, should definitely be done. Even if they will be filled in soon, I don't think incomplete functions should be merged into the main branch unless they throw an error when called.
> 
tropical_linear_space?
or is this a type?
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/1152#discussion_r827962685
> You are receiving this because you commented.
> 
> Message ID: ***@***.***>

fingolfin commented on src/TropicalGeometry/groebner_basis.jl:
```diff
@@ -0,0 +1,129 @@
+For the definitions of initial form, initial ideal and Groebner basis see
+Section 2.4 of [MS15](@cite).
+
+# Warning
```
Consider using an [admonition](https://docs.julialang.org/en/v1/stdlib/Markdown/#Admonitions) like `!!! warning` for cool effect (details in the docs I linked)

lkastner committed:
TropicalGeometry: Remove permanent

lkastner committed:
TropicalGeometry: Use admonition for warning

YueRen committed:
TropicalGeometry: docstring for tropical_semiring

YueRen committed:
TropicalGeometry: det for Matrix{Oscar.TropicalSemiringElem{T}}

YueRen committed:
TropicalGeometry: added example for det in doc of tropical_semiring

YueRen committed:
TropicalGeometry: tropical_semiring, changed warning in docstring

YueRen committed:
TropicalGeometry: _is_homogeneous, fixed typo in documentation

YueRen committed:
TropicalGeometry: _has_homogeneous_generators, made function internal + added short documentation

YueRen committed:
TropicalGeometry: _x_monomial_lt, made function internal + added short documentation

YueRen committed:
TropicalGeometry: groebner_basis, added check that generators of I are homogeneous to avoid crash if they are not

YueRen committed:
TropicalGeometry: groebner_basis, changed docstring

fingolfin commented on src/TropicalGeometry/groebner_basis.jl:
```diff
@@ -0,0 +1,130 @@
+
+
+@doc Markdown.doc"""
+    groebner_basis(I::Ideal, val::TropicalSemiringMap, w::Vector; complete_reduction::Bool, return_lead)
```
```suggestion
    groebner_basis(I::Ideal, val::TropicalSemiringMap, w::Vector; complete_reduction::Bool, return_lead::Bool)
```

fingolfin commented on src/TropicalGeometry/hypersurface.jl:
```diff
@@ -0,0 +1,260 @@
+###
+
+@doc Markdown.doc"""
+    TropicalHypersurface(f::MPolyElem{TropicalSemiringElem})
```
Again, this should match the function signature. But maybe that signature also should be adjusted...

fingolfin commented on src/TropicalGeometry/hypersurface.jl:
```diff
@@ -0,0 +1,260 @@
+function TropicalHypersurface(
+    f::Union{AbstractAlgebra.Generic.MPoly{Oscar.TropicalSemiringElem{typeof(min)}},
+             AbstractAlgebra.Generic.MPoly{Oscar.TropicalSemiringElem{typeof(max)}}}
+        )
```
How about this (more compact, easier to read)
```suggestion
    f::AbstractAlgebra.Generic.MPoly{Oscar.TropicalSemiringElem{T}}) where T
```

fingolfin commented on src/TropicalGeometry/hypersurface.jl:
```diff
@@ -0,0 +1,260 @@
+If coefficient ring has no valuation, the tropical hypersurface will be constructed with respect to the trivial valuation.
+
+# Examples
+julia> K = PadicField(7, 2)
```
This is not inside a code bloc, i.e. the triple backticks are missing

fingolfin commented on src/TropicalGeometry/hypersurface.jl:
```diff
@@ -0,0 +1,260 @@
+    set_attribute!(Tf,:algebraic_polynomial,f)
+    set_attribute!(Tf,:tropical_polynomial,tropf)
+    set_attribute!(Tf,:weights,w)
+  return Tf
```
Inconsistent indentation

YueRen committed:
TropicalGeometry: TropicalSemiringMap, expanded docstring

Commit:
TropicalGeometry: changes in Hypersurface.jl based on Max's suggestions

YueRen committed:
TropicalGeometry: tropical_polynomial, added + changed docstring

YueRen committed:
TropicalGeometry: convention, added docstring + moved export to top

YueRen committed:
TropicalGeometry: tropical_polynomial, changed input from AbstractAlgebra.Generic.MPoly to MPolyElem

YueRen committed:
TropicalGeometry: TropicalHypersurface, changed input from AbstractAlgebra.Generic.MPoly to MPolyElem

lkastner committed:
TropicalGeometry: Fix jldoctest's

lkastner committed:
TropicalGeometry: Remove superflous header line

lkastner committed:
TropicalGeometry: singular_ring->singular_coeff_ring

YueRen committed:
TropicalGeometry: tropical_semiring, renamed to TropicalSemiring

YueRen committed:
TropicalGeometry: tropical_polynomial, changed docstring

YueRen committed:
TropicalGeometry: TropicalSemiring, changed docstring

YueRen committed:
TropicalGeometry: TropicalSemiringMap, changed docstring

YueRen committed:
TropicalGeometry: bib, added Joswig 'Essentials of tropical combinatorics'

YueRen committed:
TropicalGeometry: TropicalSemiringMap, changed docstring

YueRen committed:
TropicalGeometry: intro, overhauled

lkastner committed:
Update src/TropicalGeometry/groebner_basis.jl

Co-authored-by: Max Horn <max@quendi.de>

fieker commented:
You don't need the solver... Yue has a replacement within Oscar for it ..


On Wed, 9 Mar 2022, 11:03 Lars Kastner, ***@***.***> wrote:

> This is a subset of the big tropical geometry branch which contains the
> basic architecture, but does not need the pAdicSolver. So this can be
> merged and then some other branches will be rebased onto this.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/oscar-system/Oscar.jl/pull/1152
> Commit Summary
>
>    - 74bc4fd
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/74bc4fd09c0a9355ea35c05c08ab5a41fbf6e228>
>    More fixes to satisfy the incoming MapFromFunc changes
>    - e521ff5
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/e521ff5e14f01dc66a3f666d6aa6702cc41605e0>
>    PolyhedralGeometry: Access to basic properties of PolyhedralComplex
>    - 87b9c4f
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/87b9c4f91f2bba5b82aa6afc77f7081eadfd4808>
>    PolyhedralGeometry: Add some more basic functionality for PolyhedralComplex
>    - 4bf592e
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/4bf592eb66cfa220eeaa665caa069e7c617aa976>
>    TropicalGeometry: initial commit
>    - 229bdf4
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/229bdf4db4b8fe4513be0410d0c292d898413bb3>
>    TropicalGeometry: Make TropicalHypersurface(poly) work
>    - a66cdf2
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/a66cdf26c620b2647b0fc19fb9a6dcfe6add5984>
>    TropicalGeometry: Homogenize inhomogeneous input for polymake
>    - f860efc
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/f860efc07b2d1d22c975afd778a73d240d838fb2>
>    TropicalGeomtry: Ask for vertices of tropical hypersurface
>    - 81be961
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/81be961b17d6884c0c2054a66818bd94f2f01dfe>
>    chg: moving files around, adding basic properties for tropical varieties
>    - 77d6116
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/77d61164fd166e0c9a9f69f3f0d35431f55a3ce7>
>    fix: some broken properties
>    - c9fcbf4
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/c9fcbf4bed809b93889e10826180f1196498872b>
>    fix: broken property
>    - 2687e87
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/2687e87e2b0e7a8f97f810fe5947f3c4b301c44d>
>    TropicalGeometry: Also homogenize homogeneous polynomials, i.e. add variable
>    - 2958f20
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/2958f206e1c9e40ee7474619dcfe1ee790cc6990>
>    fix: tropical numbers
>    - 0d35642
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/0d356424f3be3a41789569c350e218ed64d7ce7c>
>    TropicalGeometry: Remove brainstorming leftovers
>    - 17d8589
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/17d8589529da9f68c3dbd36495074199fc8f67d3>
>    TropicalGeometry: Nicer show methods
>    - bdbacd0
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/bdbacd047260d34f1036e8f43026d5c6b67ee78b>
>    TropicalGeometry: intersect tropical varieties
>    - 22887a1
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/22887a1c8877af733172f333c060c3a10d4cc999>
>    TropicalGeometry: Add documentation to intersect function
>    - 07fa000
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/07fa000a142b5f99e873b9446e9af565e2d96c48>
>    TropicalGeometry: Add stably_intersect
>    - 5105d9b
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/5105d9bf14a589dd9916edb85e88261397f16472>
>    TropicalGeometry: Fix docstring for new Base.show
>    - 5c063a5
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/5c063a55d702594ae6ac1a19580af646b143df22>
>    TropicalGeometry: Eliminate most warnings from duplicate definitions
>    - 263ed8b
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/263ed8b79ff53e10778ab038583adca8648c215d>
>    chg (broken): restructured codebase
>    - 4aa2595
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/4aa259572cc777bb33774164da41912cf887c692>
>    TropicalGeometry: Add constructor according to Note in AbstractAlgebra.jl
>    - be3ceab
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/be3ceab37031ed05e4883583954251fd4aa4ad5f>
>    TropicalGeometry: Start documentation
>    - 54ec8ca
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/54ec8caa1fafa305dd0c35534339c6505ba407bd>
>    chg (broken): constructors for tropical hypersurfaces
>    - 74adf4d
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/74adf4d94839ec244c1e187f90ac7a1c7ccfee04>
>    fix: hypersurface from algebraic polynomial
>    - ca5c3f8
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/ca5c3f8b9d89dfacb102276a81d75c25c47496f7>
>    TropicalGeometry: Extract underlying poyhedral complex
>    - 1cf2edc
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/1cf2edc30356c277b557bde892173b968b7c517a>
>    fix: tropical_polynomial coefficients depend on min-max convention
>    - 13d515d
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/13d515d35b80b027d400d75fd352ba20cc0f1d6d>
>    chg: adding more documentation
>    - 6477fa2
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/6477fa2a5a58c94974a0a2d0adb871535ccc5bb5>
>    chg: initial changes to base everything on PolyhedralComplex
>    - 5c9ec95
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/5c9ec95c78c1433cddfe460a2a8e240ec5a9e667>
>    fix: various bugs
>    - 93a665e
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/93a665e25e0ec5c0d56a104c033a8ad6620a46aa>
>    TropicalGeometry: initial commit for valuation.jl
>    - 8f46ba0
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/8f46ba06526ddea88c8a2e7f47a487b197a1aef6>
>    TropicalGeometry: fixing initial forms to work with ValuationMap
>    - 67c39ba
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/67c39bab14da0f0d801e4ca2f1535a0df0b63918>
>    TropicalGeometry: support for trivial valuations
>    - fabacec
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/fabacec475a8e3e3a22ea3e4087d5ca1a93f02c5>
>    TropicalGeometry: initial commit for initial.jl
>    - 07bdad6
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/07bdad644ab74455784969f540eb00b9546edbf4>
>    TropicalGeometry: initial commit for groebner_basis.jl
>    - f80bc07
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/f80bc077bde933ea124149188a1e22bfe784d9f6>
>    TropicalGeometry: initial (broken) version for tropical groebner bases
>    - 15dcc76
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/15dcc767a3e69453cd92f400ae703fa3ef8d2398>
>    TropicalGeometry: fixing tropical groebner bases
>    - bb6cae2
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/bb6cae286b8a537442e12d7d910aa75f429a4ec0>
>    TropicalGeometry: fix groebner_basis header
>    - 078d2f5
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/078d2f5d00857f71c80428bf63596334941aa177>
>    TropicalGeometry: sanatizing input and names in groebner_bases.jl
>    - c350492
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/c3504922447433aa3b6f8250be427d13e9be4b85>
>    TropicalGeometry: adding complete_reduction parameter to groebner basis
>    - 12e2e05
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/12e2e056a3bbf8add39ce7989c506f08c453cdeb>
>    TropicalGeometry: using MPolyBuildCtx in pseudo_change_base_ring
>    - b151b02
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/b151b02240872498f77a115d0eb3b23957bbf9a7>
>    TropicalGeometry: desimulate_valuation
>    - baf517a
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/baf517a41c0d0d76a86266648e5c0f280078de6d>
>    TropicalGeometry: code cleanup
>    - 2f46346
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/2f46346ac8c7b4281dabe386d80eb465c409e7c1>
>    TropicalGeometry: added return_lead parameter to groebner_basis
>    - 516a5f2
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/516a5f2e0f7cc22d66d23996da099acb773bcf67>
>    TropicalGeometry: initial commit groebner_polyhedron.jl
>    - 1b0de97
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/1b0de97adcb3364ab48d77ec9f5a41f65e8bda4d>
>    TropicalGeometry: new homogeneity_space
>    - 7ffe037
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/7ffe03747570e34e9403db9d63c14a8afb607527>
>    TropicalGeometry: code cleanup using Iterators.peel
>    - f49e4cb
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/f49e4cba36e557d861f215cd602e5a6d06299849>
>    TropicalGeometry: initial commit for points.jl, added
>    random_linear_polynomials
>    - 161fc48
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/161fc48e7f142f8b3788b00fec4bf3fdbf2fca4a>
>    TropicalGeometry: broken commit so we can continue working on Avi's desktop
>    - 62cbd1d
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/62cbd1d23615aa85544460f5f3c0349de0912cb8>
>    TropicalGeometry: fixing random_linear_polynomials to return non-homogeous
>    polynomials
>    - a121f69
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/a121f69b8db4bb33f544f06d5e4c1fb22ee352cb>
>    TropicalGeometry: renamed random_linear_polynomial to
>    random_affine_linear_polynomial
>    - 7e5b9b3
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/7e5b9b3890d4b5ec653fbc59a0722b6cbf816364>
>    TropicalGeometry: first version of tropical_link
>    - 6ba3d43
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/6ba3d43941e0899604883da4f57a48b7feb41f9f>
>    TropicalGeometry: fixed tropical_link to ignore multiplicities
>    - 99c84f7
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/99c84f7027ca431d3bcc919897592571b2235225>
>    TropicalGeometry: typos and redundant Oscar.
>    - 835beff
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/835beffd6bcb597a92504672c2a6d249c1ed10d6>
>    TropicalGeometry: fixed check_weight_vector_for_legality
>    - 18d36f5
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/18d36f542ce2cc4907a4224a11532b2e7c35947f>
>    TropicalGeometry: code cleanup
>    - c716a39
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/c716a3951837d36bbc543cedc7601a487b642ffb>
>    TropicalGeometry: added desimulate_valuation for weight vectors
>    - a974ad3
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/a974ad39c2f1365149dd4464c9b32adfb61ad63b>
>    TropicalGeometry: fixing valuations + initial for ideals
>    - 8a91613
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/8a91613478488bc189c491f9077cd4f78bc5b8ef>
>    TropicalGeometry: major ValuationMap overhaul
>    - 65b2e67
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/65b2e67111a4346194575be38b3da296dc66e25e>
>    TropicalGeometry: added docu for groebner_basis
>    - 3eda357
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/3eda357cd92ec37e170b6d5b7574ef267d5cd46b>
>    TropicalGeometry: added weight for trop hypersurfaces
>    - 04b7277
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/04b7277feeafedf92c8b2ebb3ddff65975bfdb2e>
>    TropicalGeometry: added desimulate_valuation for pairs of vectors
>    - 4837c67
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/4837c67ba537d7feecc34896521fbde3db44dcd1>
>    TropicalGeometry: fixed computing tropical reduced GB
>    - 0ead060
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/0ead0606f94e6397566e76550f9041c5db23fe49>
>    TropicalGeometry: fixed tighten_simulation to create distinct x-monomials
>    - 210ae3d
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/210ae3d9c2b7a5520a6c8130f8af1e8a1038010b>
>    TropicalGeometry: added show functions for ValuationMap
>    - a1a2ef2
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/a1a2ef2610d1597b34a9e4874c1567fa9b1dfc42>
>    TropicalGeometry: tighten_simulation fixed to divide by gcd of coefficients
>    - 81cf125
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/81cf125e5da0bdb774b5e634a59a1b63f0accf06>
>    TropicalGeometry: simulate_valuation restructured to allow for vectors of
>    polys as input
>    - af5d7bf
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/af5d7bfc2248737059ea979243281c9dc08349f3>
>    TropicalGeometry: is_valuation_*, making tests more flexible
>    - 99ffd51
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/99ffd51d50cb781bbe552dff0533a5cd66ebe4dc>
>    TropicalGeometry: tighten_valuation, improved docu
>    - 085bb7b
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/085bb7b0c7a15bb2986e4cb31b97b9179dbec7e2>
>    TropicalGeometry: t_adic_valuation, extended to more input types
>    - 82a255a
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/82a255abfffe617a6ae797cecec5e23e5aa03516>
>    TropicalGeometry: TropicalHypersurface, default to min and allow passing
>    valuation as argument
>    - dfbbb1b
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/dfbbb1b6cca9f04895b9e94a3f523d7aac9f8972>
>    TropicalGeometry: renamed tropical_numbers to tropical_semiring
>    - a0a0aee
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/commits/a0a0aee805108b378739662377ddb855f0396a02>
>    TropicalGeometry: initial restructured to allow polys, vectors and ideals
>    as input
>
> File Changes
>
> (22 files <https://github.com/oscar-system/Oscar.jl/pull/1152/files>)
>
>    - *M* docs/doc.main
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-25080aaf5efd96a8fd7f585fcee40bc1437001ada394d070e716ecc10786a55b>
>    (3)
>    - *M* docs/oscar_references.bib
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-073962c13ac44fe9a0b503c792b850ea09c2065492151fb16e364867c29dc4cc>
>    (13)
>    - *A* docs/src/TropicalGeometry/intro.md
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-b044e7489afc7c5e204a5ffc28ebf1bbc370d6b64e547bb3639e9bb7f21e4d88>
>    (29)
>    - *M* experimental/GModule/Cohomology.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-79f3f094f8fabbd9f1dac92c0eac7be2f5a9505a5f48a79cb519943dc5310dff>
>    (2)
>    - *M* experimental/GModule/GModule.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-d52f8a20b9ef93c33fc9a5eab5d5d675c849c624cac618ec9323f2900c048059>
>    (3)
>    - *M* src/Modules/ModuleTypes.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-c5b66874d96827632399fa60ef2e97345d7e9851ed184c6ecb338fe6c26be0ba>
>    (6)
>    - *M* src/Oscar.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-2b0f42aa9812c1d3a0e1613df08cd89260655b4016bd4bebad32e3037bf1e2be>
>    (1)
>    - *A* src/TropicalGeometry/curve.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-039892c0054c6d12760526ef4f950c422410d31987dbd9c6ebc36b1ee3955b37>
>    (67)
>    - *A* src/TropicalGeometry/groebner_basis.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-b391016fc4d30d27bde3ee2a0334e8a0d281aecb6366a240d0aec5cbdbbb8100>
>    (171)
>    - *A* src/TropicalGeometry/groebner_polyhedron.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-846e1ff604d27d15ce7ce884839ce9ed60997903a1c5d3c8a680305b4dedc5c0>
>    (48)
>    - *A* src/TropicalGeometry/hypersurface.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-d12016368b54900c809d98ff9633a4db3cbd47e66cf54b548aa1d2051cb4d1a6>
>    (256)
>    - *A* src/TropicalGeometry/initial.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-e89c52fb72b9478893aabb99f0b7c857f91607efabc0e07be2f10ce0137540d3>
>    (194)
>    - *A* src/TropicalGeometry/linear_space.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-b089bb67cb183bde5e1a8c843bee1ab63b2fddd850f38aeb26f94951d2f280b9>
>    (60)
>    - *A* src/TropicalGeometry/links.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-3b8a1c06a8c81526339121eea4af0cf365c72a8abd904c5cd05f840d3752089a>
>    (172)
>    - *A* src/TropicalGeometry/main.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-5daae5fb6ce745fcdf9067da9a19945f89f7c6390e9900d6fa4e2aab02994b10>
>    (80)
>    - *A* src/TropicalGeometry/numbers.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-a78125a04706720bb3becd835c7c46016c27899c783ed52badd7d6eb92fdc28c>
>    (449)
>    - *A* src/TropicalGeometry/points.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-f544a33c42eef58cf9a2719c343c22d1dafe71ddb6a5c512680a64f9b966b433>
>    (77)
>    - *A* src/TropicalGeometry/poly.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-61b152f79195a0db54245c30b2deec6ec14a3cef831caac84701db580a27c3ce>
>    (55)
>    - *A* src/TropicalGeometry/semiring.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-a6b6c221482d03057ce09f1d0eb458e6c386065671c29bb083aef639d12f92bc>
>    (504)
>    - *A* src/TropicalGeometry/valuation.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-c9638203ced3d0a956212cfc0d9438bad23b7ba587c85161c47a0fb12878159e>
>    (442)
>    - *A* src/TropicalGeometry/variety.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-0379b0f2b166f68b219d831526ab1e7c3e0894abb0f5db4c14b1cadb509e0ed1>
>    (93)
>    - *A* src/TropicalGeometry/variety_supertype.jl
>    <https://github.com/oscar-system/Oscar.jl/pull/1152/files#diff-19ad7969d831969c44e0286d2ca9a718ae16170630d0fc2a4b6729753186eb37>
>    (590)
>
> Patch Links:
>
>    - https://github.com/oscar-system/Oscar.jl/pull/1152.patch
>    - https://github.com/oscar-system/Oscar.jl/pull/1152.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/oscar-system/Oscar.jl/pull/1152>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA36CVYPS7PKN4S5GDSSSYTU7BZO3ANCNFSM5QJBGAQQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
